File: mcp-code-intelligence-nodejs/src/orchestration/local/process.ts
/**
 * Single child MCP server process lifecycle — spawn, initialize, fetch tools, health check.
 * State machine: STARTING → READY → ACTIVE → CRASHED → RESTARTING → DEAD.
 * Behavioral parity with Kotlin ServerProcess.kt.
 */

import { spawn, ChildProcess, execSync } from 'child_process';
import { ServerEntry } from '../config.js';
import { StdioJsonRpc } from './rpc.js';

export enum ServerState {
  STARTING = 'STARTING', READY = 'READY', ACTIVE = 'ACTIVE',
  CRASHED = 'CRASHED', RESTARTING = 'RESTARTING',
  STOPPING = 'STOPPING', DEAD = 'DEAD', FAILED = 'FAILED',
}

export class ServerProcess {
  readonly name: string;
  state: ServerState = ServerState.STARTING;
  tools: Record<string, any>[] = [];
  retryCount = 0;
  readonly rpc = new StdioJsonRpc();
  private proc: ChildProcess | null = null;
  private entry: ServerEntry;

  constructor(name: string, entry: ServerEntry) {
    this.name = name;
    this.entry = entry;
  }

  /** Start child process, initialize MCP handshake, fetch tools. */
  async start(): Promise<boolean> {
    this.state = ServerState.STARTING;
    const proc = this.spawnProcess();
    if (!proc) return this.markFailed('Failed to spawn process');
    this.proc = proc;
    this.rpc.attach(proc);
    if (!await this.initialize()) return this.markFailed('Initialize handshake failed');
    this.state = ServerState.READY;
    if (!await this.fetchTools()) return this.markFailed('Failed to fetch tools');
    this.state = ServerState.ACTIVE;
    this.log(`Active with ${this.tools.length} tools`);
    return true;
  }

  /** Stop child process gracefully. */
  stop(): void {
    this.state = ServerState.STOPPING;
    this.rpc.detach();
    this.destroyProcess();
    this.state = ServerState.DEAD;
    this.log('Stopped');
  }

  /** Restart after crash with exponential backoff. */
  async restart(maxRetries: number): Promise<boolean> {
    if (this.retryCount >= maxRetries) { this.state = ServerState.DEAD; return false; }
    this.state = ServerState.RESTARTING;
    this.retryCount++;
    const backoffMs = Math.min(1000 * this.retryCount, 10_000);
    this.log(`Restarting (attempt ${this.retryCount}/${maxRetries}, backoff ${backoffMs}ms)`);
    await new Promise((r) => setTimeout(r, backoffMs));
    this.destroyProcess();
    return this.start();
  }

  /** Call a tool on this child server via JSON-RPC. */
  async callTool(toolName: string, args: Record<string, any>, timeoutMs: number): Promise<any> {
    return this.rpc.sendRequest('tools/call', { name: toolName, arguments: args }, timeoutMs);
  }

  /** Health check — send tools/list, expect response within 5s. */
  async healthCheck(): Promise<boolean> {
    try { await this.rpc.sendRequest('tools/list', {}, 5_000); return true; }
    catch { return false; }
  }

  /** Check if OS process is still alive. */
  isAlive(): boolean { return this.proc !== null && this.proc.exitCode === null; }

  private spawnProcess(): ChildProcess | null {
    try {
      const isWin = process.platform === 'win32';
      return spawn(this.entry.command, this.entry.args, {
        stdio: ['pipe', 'pipe', 'pipe'],
        env: { ...process.env, ...this.entry.env },
        shell: isWin,
      });
    } catch (e: any) {
      this.log(`Spawn failed: ${e.message}`);
      return null;
    }
  }

  private async initialize(): Promise<boolean> {
    const params = {
      protocolVersion: '2024-11-05',
      capabilities: {},
      clientInfo: { name: 'mcp-orchestrator', version: '1.0.0' },
    };
    try {
      await this.rpc.sendRequest('initialize', params, this.entry.timeout);
      this.rpc.sendNotification('notifications/initialized', {});
      return true;
    } catch (e: any) {
      this.log(`Initialize failed: ${e.message}`);
      return false;
    }
  }

  private async fetchTools(): Promise<boolean> {
    try {
      const result = await this.rpc.sendRequest('tools/list', {}, this.entry.timeout);
      this.tools = result?.tools ?? [];
      return true;
    } catch (e: any) {
      this.log(`Fetch tools failed: ${e.message}`);
      return false;
    }
  }

  private destroyProcess(): void {
    const proc = this.proc;
    if (!proc) return;
    if (proc.exitCode === null) {
      if (!this.tryWindowsTreeKill(proc)) proc.kill('SIGKILL');
    }
    this.proc = null;
  }

  private tryWindowsTreeKill(proc: ChildProcess): boolean {
    if (process.platform !== 'win32') return false;
    try {
      execSync(`taskkill /T /F /PID ${proc.pid}`, { timeout: 3000 });
      return true;
    } catch { return false; }
  }

  private markFailed(reason: string): boolean {
    this.log(reason);
    this.state = ServerState.FAILED;
    return false;
  }

  private log(msg: string): void { console.error(`[${this.name}] ${msg}`); }
}

File: mcp-code-intelligence-python/src/mcp_code_intel/orchestration/local/process.py
"""Single child MCP server process lifecycle — spawn, initialize, fetch tools, health check.

State machine: STARTING → READY → ACTIVE → CRASHED → RESTARTING → DEAD.
Behavioral parity with Kotlin ServerProcess.kt.
"""

from __future__ import annotations

import asyncio
import os
import platform
import subprocess
import sys
from enum import Enum
from typing import Any

from ..config import ServerEntry
from .rpc import StdioJsonRpc


class ServerState(Enum):
    """Server lifecycle states."""

    STARTING = "STARTING"
    READY = "READY"
    ACTIVE = "ACTIVE"
    CRASHED = "CRASHED"
    RESTARTING = "RESTARTING"
    STOPPING = "STOPPING"
    DEAD = "DEAD"
    FAILED = "FAILED"


class ServerProcess:
    """Manages a single child MCP server process."""

    def __init__(self, name: str, entry: ServerEntry) -> None:
        self.name = name
        self.state = ServerState.STARTING
        self.tools: list[dict] = []
        self.retry_count = 0
        self._entry = entry
        self._proc: asyncio.subprocess.Process | None = None
        self.rpc = StdioJsonRpc()

    async def start(self) -> bool:
        """Start child process, initialize MCP handshake, fetch tools."""
        self.state = ServerState.STARTING
        proc = await self._spawn()
        if not proc:
            return self._mark_failed("Failed to spawn process")
        self._proc = proc
        self.rpc.attach(proc)
        if not await self._initialize():
            return self._mark_failed("Initialize handshake failed")
        self.state = ServerState.READY
        if not await self._fetch_tools():
            return self._mark_failed("Failed to fetch tools")
        self.state = ServerState.ACTIVE
        self._log(f"Active with {len(self.tools)} tools")
        return True

    def stop(self) -> None:
        """Stop child process gracefully."""
        self.state = ServerState.STOPPING
        self.rpc.detach()
        self._destroy_process()
        self.state = ServerState.DEAD
        self._log("Stopped")

    async def restart(self, max_retries: int) -> bool:
        """Restart after crash with exponential backoff."""
        if self.retry_count >= max_retries:
            self.state = ServerState.DEAD
            return False
        self.state = ServerState.RESTARTING
        self.retry_count += 1
        backoff_s = min(1.0 * self.retry_count, 10.0)
        self._log(f"Restarting (attempt {self.retry_count}/{max_retries}, backoff {backoff_s}s)")
        await asyncio.sleep(backoff_s)
        self._destroy_process()
        return await self.start()

    async def call_tool(self, tool_name: str, args: dict, timeout_ms: int) -> Any:
        """Call a tool on this child server via JSON-RPC."""
        params = {"name": tool_name, "arguments": args}
        return await self.rpc.send_request("tools/call", params, timeout_ms)

    async def health_check(self) -> bool:
        """Health check — send tools/list, expect response within 5s."""
        try:
            await self.rpc.send_request("tools/list", {}, 5_000)
            return True
        except Exception:
            return False

    def is_alive(self) -> bool:
        """Check if OS process is still running."""
        return self._proc is not None and self._proc.returncode is None

    async def _spawn(self) -> asyncio.subprocess.Process | None:
        """Spawn child process with configured command and args."""
        try:
            if not self._entry.command:
                self._log("No command specified")
                return None
            cmd = self._resolve_command(self._entry.command)
            args = list(self._entry.args)
            # Note: depth args NOT injected for third-party servers
            # (they don't understand --depth/--max-depth)
            env = {**os.environ, **self._entry.env}
            return await asyncio.create_subprocess_exec(
                *cmd, *args,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                env=env,
            )
        except Exception as e:
            self._log(f"Spawn failed: {e}")
            return None

    def _resolve_command(self, cmd: str) -> list[str]:
        """On Windows, wrap non-exe commands with cmd /c."""
        if not self._is_windows():
            return [cmd]
        lower = cmd.lower()
        if lower.endswith((".exe", ".bat", ".cmd")):
            return [cmd]
        return ["cmd", "/c", cmd]

    async def _initialize(self) -> bool:
        """Send MCP initialize handshake."""
        params = {
            "protocolVersion": "2024-11-05",
            "capabilities": {},
            "clientInfo": {"name": "mcp-orchestrator", "version": "1.0.0"},
        }
        try:
            await self.rpc.send_request("initialize", params, self._entry.timeout)
            self.rpc.send_notification("notifications/initialized", {})
            return True
        except Exception as e:
            self._log(f"Initialize failed: {e}")
            return False

    async def _fetch_tools(self) -> bool:
        """Fetch tools/list from child server."""
        try:
            result = await self.rpc.send_request("tools/list", {}, self._entry.timeout)
            self.tools = result.get("tools", []) if isinstance(result, dict) else []
            return True
        except Exception as e:
            self._log(f"Fetch tools failed: {e}")
            return False

    def _destroy_process(self) -> None:
        """Kill child process (tree kill on Windows)."""
        proc = self._proc
        if not proc:
            return
        if proc.returncode is None:
            if not self._try_windows_tree_kill(proc):
                proc.kill()
        self._proc = None

    def _try_windows_tree_kill(self, proc: asyncio.subprocess.Process) -> bool:
        """On Windows, use taskkill /T /F /PID for tree kill."""
        if not self._is_windows():
            return False
        try:
            pid = proc.pid
            subprocess.run(
                ["taskkill", "/T", "/F", "/PID", str(pid)],
                capture_output=True, timeout=3,
            )
            return True
        except Exception:
            return False

    def _is_windows(self) -> bool:
        return platform.system() == "Windows"

    def _mark_failed(self, reason: str) -> bool:
        self._log(reason)
        self.state = ServerState.FAILED
        return False

    def _log(self, msg: str) -> None:
        print(f"[{self.name}] {msg}", file=sys.stderr, flush=True)
